kernel/copy.masm

```diff
@@ -0,0 +1,256 @@
+# `writePtr` address, the hash of the copied data is left on the stack for
+# commitment checking.
+proc.move_from_tape_to_memory.0
+  # input: [words, writePtr, ...]
+  # output: [A, B, C, writePtr', ...]
+  
+  # check if there is an odd number of words
+  # stack: [is_odd, words, writePtr, ...]
+  dup exec.is_odd
+
+  # copy the flag `is_odd`, this will later on define if padding is required
+  # stack: [is_odd, words, writePtr, needs_padding, ...]
+  dup movdn.3
+
+  # Make sure `words` is event. This is a requirement for the
+  # `move_from_tape_to_memory_even_words`.
+  # - even words will be `words-0` so result stays even
+  # - odd words will be `words-1` so result becomes event
+  #
+  # stack: [words_even, writePtr, needs_padding, ...]
+  sub
+
+  # Prepare the capacity word. For rescue prime optimized the first element is
+  # set to `1` when padding is used and `0` otherwse, this is determined by the
+  # `needs_padding` flag, so just copy it.
+  dup.1 push.0.0.0
+
+  # set initial state
+  # stack: [A, B, C, writePtr, words_even, needs_padding, ...]
+  padw padw
+
+  # stack: [A, B, C, writePtr, needs_padding, ...]
+  exec.move_from_tape_to_memory_even_words
+
+  # stack: [needs_padding, A, B, C, writePtr, ...]
+  movup.13
+
+  # if(needs_padding)
+  if.true 
+    # Rescue Prime Optimized uses overwrite mode. Drop `A`.
+    # stack: [B, C, writePtr, ...]
+    dropw
+
+    # Overwrite the `B` with the new data
+    # stack: [B', C, writePtr, ...]
+    adv_loadw
+
+    # - get the memory address that B' should be saved to
+    # - update the writePtr to point to the next address
+    # stack: [writePtr, B', writePtr+1, C, ...]
+    movup.8 dup.0 add.1 movdn.5
+
+    # save data to memory
+    # stack: [B', writePtr+1, C, ...]
+    mem_storew
+
+    # Fix writePtr position
+    # stack: [B', C, writePtr+1, ...]
+    movup.4 movdn.8
+
+    # Push padding word
+    # stack: [A, B', C, writePtr+1, ...]
+    push.1.0.0.0
+
+    # Run RPO permutation
+    # stack: [A', B', C', writePtr+1, ...]
+    rpperm
+  end
+
+  # The RPO hash is a single word, discard the unused one and the capacity.
+  # stack: [RpoHash, writePtr', ...]
+  dropw swapw dropw
+end
+
+# move_from_tape_to_memory_with_commitment(words: Felt, writePtr: Address, commitment: Commitment) -> [writePtr']
+#
+# Moves the `words` words from the advice tape to memory, starting at
+# `writePtr` address, at the end asserts the commitment data was copied
+proc.move_from_tape_to_memory_with_commitment.0
+  # Copies the tape data to memory
+  # stack: [RpoHash, writePtr', Commitment, ...]
+  exec.move_from_tape_to_memory
+
+  # Save the writePtr
+  # stack: [RpoHash, Commitment, writePtr', ...]
+  movup.4 movdn.8
+
+  # Check the commitment
+  # stack: [writePtr', ...]
+  exec.assert_eqw
+end
+
+# memcopy(readPtr: Address, writePtr: Address, words: Felt)
+#
+# Copies `words` words from `readPtr` to `writePtr`.
+proc.memcopy.0
```

Comment:
This procedure could probably go into the standard library under `std::mem::memcopy`. In the future, we might even introduce an instruction to perform this in a single VM cycle.

Reply:
That is one thing that I had in mind while writing these procedures. I had two goals:

1. write reusable procedures
2. make it easy to change our format (add/remove fields), That is why everything is done with a loop, instead of a manually unrolled loop.

Let me know if I should move these to a PR to `miden-vm`.

Reply:
Yes, let's make a PR against `miden-vm`, but only for this procedure for now.

Reply:
PR is open: https://github.com/0xPolygonMiden/miden-vm/pull/668